spynnaker/pyNN/models/neuron/synaptic_matrix.py

```diff
@@ -0,0 +1,585 @@
+        :param MaxRowInfo max_row_info: Maximum row length information
+        :param PartitionRoutingInfo routing_info:
+            Routing information for the edge
+        :param PartitionRoutingInfo delay_routing_info:
+            Routing information for the delay edge if any
+        :param list(float) weight_scales: Weight scale for each synapse type
+        :param all_syn_block_sz:
+            The space available for all synaptic matrices
+        :param int all_single_syn_sz:
+            The space available for "direct" or "single" synapses
+        """
+        self.__synapse_io = synapse_io
+        self.__poptable = poptable
+        self.__synapse_info = synapse_info
+        self.__machine_edge = machine_edge
+        self.__app_edge = app_edge
+        self.__n_synapse_types = n_synapse_types
+        self.__max_row_info = max_row_info
+        self.__routing_info = routing_info
+        self.__delay_routing_info = delay_routing_info
+        self.__weight_scales = weight_scales
+        self.__all_syn_block_sz = all_syn_block_sz
+        self.__all_single_syn_sz = all_single_syn_sz
+
+        # The matrix size can be calculated up-front; use for checking later
+        self.__matrix_size = (
+            self.__max_row_info.undelayed_max_bytes *
+            self.__machine_edge.pre_vertex.vertex_slice.n_atoms)
+        self.__delay_matrix_size = (
+            self.__max_row_info.delayed_max_bytes *
+            self.__machine_edge.pre_vertex.vertex_slice.n_atoms *
+            self.__app_edge.n_delay_stages)
+        self.__single_matrix_size = (
+            self.__machine_edge.pre_vertex.vertex_slice.n_atoms *
+            BYTES_PER_WORD)
+
+        self.__index = None
+        self.__delay_index = None
+        self.__syn_mat_offset = None
+        self.__delay_syn_mat_offset = None
+        self.__is_single = False
+        self.__received_block = None
+        self.__delay_received_block = None
+
+    @property
+    def is_delayed(self):
+        """ Is there a delay matrix?
+
+        :rtype: bool
+        """
+        return self.__app_edge.n_delay_stages > 0
+
+    def is_direct(self, single_addr):
+        """ Determine if the given connection can be done with a "direct"\
+            synaptic matrix - this must have an exactly 1 entry per row
+
+        :param int single_addr: The current offset of the direct matrix
+        :return: A tuple of a boolean indicating if the matrix is direct and
+            the next offset of the single matrix
+        :rtype: (bool, int)
+        """
+        pre_vertex_slice = self.__machine_edge.pre_vertex.vertex_slice
+        post_vertex_slice = self.__machine_edge.post_vertex.vertex_slice
+        next_addr = single_addr + self.__single_matrix_size
+        is_direct = (
+            next_addr <= self.__all_single_syn_sz and
+            not self.is_delayed and
+            isinstance(self.__synapse_info.connector, OneToOneConnector) and
```

Comment:
If we had the minimum and maximum row length of the actual data, and both were equal to 1, the row can use the direct synapses.  However this was for a specific use case (one-to-one Poisson sources), and is only an optimisation.  It could be done in the future if desired though.